File: intermediate_data_structures/intermediate_ds/binary_search_tree.py
class BinarySearchTree:
    class _Node:
        def __init__(self, key): self.key = key; self.left = self.right = None
    def __init__(self): self.root = None
    def insert(self, key):
        if self.root is None: self.root = self._Node(key); return
        node = self.root
        while True:
            branch = "left" if key < node.key else "right"
            child = getattr(node, branch)
            if child is None: setattr(node, branch, self._Node(key)); return
            node = child
    def contains(self, key):
        node = self.root
        while node:
            if key == node.key: return True
            node = node.left if key < node.key else node.right
        return False
    def inorder(self):
        def walk(node):
            if node: yield from walk(node.left); yield node.key; yield from walk(node.right)
        return list(walk(self.root))

File: structures/binary_search_tree.go
package structures

import "cmp"

// BinarySearchTree indexes ordered values.
type BinarySearchTree[T cmp.Ordered] struct{ root *bstNode[T] }
type bstNode[T cmp.Ordered] struct{ value T; left,right *bstNode[T] }
func(t *BinarySearchTree[T]) Insert(v T){if t.root==nil{t.root=&bstNode[T]{value:v};return};for n:=t.root;;{if v==n.value{return};if v<n.value{if n.left==nil{n.left=&bstNode[T]{value:v};return};n=n.left}else{if n.right==nil{n.right=&bstNode[T]{value:v};return};n=n.right}}}
func(t *BinarySearchTree[T]) Contains(v T)bool{for n:=t.root;n!=nil;{if n.value==v{return true};if v<n.value{n=n.left}else{n=n.right}};return false}
func(t *BinarySearchTree[T]) InOrder()[]T{out:=[]T{};var visit func(*bstNode[T]);visit=func(n *bstNode[T]){if n!=nil{visit(n.left);out=append(out,n.value);visit(n.right)}};visit(t.root);return out}
